scarlet-protocol-stomp/src/main/java/com/tinder/scarlet/stomp/okhttp/client/OkHttpStompMainChannel.kt

```diff
@@ -0,0 +1,296 @@
+                sendErrorMessage("No messages received in $interval ms.")
+                connection?.close()
+                listener.onFailed(this@OkHttpStompMainChannel, true, null)
+            }
+        }
+    }
+
+    private fun sendHeartBeat() {
+        val stompMessage = StompMessage.Builder()
+            .create(StompCommand.HEARTBEAT)
+
+        connection?.sendMessage(stompMessage)
+    }
+
+    private fun sendErrorMessage(error: String) {
+        val headers = StompHeaderAccessor.of()
+            .apply { message(error) }
+            .createHeader()
+
+        val stompMessage = StompMessage.Builder()
+            .withHeaders(headers)
+            .create(StompCommand.ERROR)
+
+        connection?.sendMessage(stompMessage)
+    }
+
+    inner class InnerWebSocketListener(
+        private val openRequest: OkHttpStompClient.ClientOpenRequest
+    ) : WebSocketListener() {
+        override fun onOpen(webSocket: WebSocket, response: Response) {
+            val webSocketConnection = WebSocketConnection(webSocket)
+
+            this@OkHttpStompMainChannel.connection = webSocketConnection
+            this@OkHttpStompMainChannel.messageHandler = webSocketConnection
+
+            val host = configuration.host
+            val login = openRequest.login
+            val passcode = openRequest.passcode
+
+            sendConnectMessage(host, login, passcode)
+        }
+
+        override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
+            messageHandler?.handle(bytes.toByteArray())?.let(::handleIncome)
+        }
+
+        override fun onMessage(webSocket: WebSocket, text: String) {
+            messageHandler?.handle(text.toByteArray(Charsets.UTF_8))?.let(::handleIncome)
+        }
+
+        override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
+            listener.onClosing(this@OkHttpStompMainChannel)
+        }
+
+        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
+            listener.onClosed(this@OkHttpStompMainChannel)
+            this@OkHttpStompMainChannel.connection = null
+        }
+
+        override fun onFailure(webSocket: WebSocket, throwable: Throwable, response: Response?) {
+            listener.onFailed(this@OkHttpStompMainChannel, true, throwable)
+            this@OkHttpStompMainChannel.connection = null
+        }
+    }
+
+    private fun sendConnectMessage(host: String, login: String? = null, passcode: String? = null) {
+        val stompHeaderAccessor = StompHeaderAccessor.of()
+            .apply {
+                this.host = host
+                this.acceptVersion =
+                    ACCEPT_VERSION
+                this.login = login
+                this.passcode = passcode
+            }
+
+        val clientSendInterval = configuration.heartbeatSendInterval
+        val clientReceiveInterval = configuration.heartbeatReceiveInterval
+
+        if (clientSendInterval > 0 && clientReceiveInterval > 0) {
```

Comment:
I would encapsulate this in `Configuration`

```kotlin
data class Configuration() {
  val shouldSendHeartBeat: Boolean
    get() = ...
}
```

Reply:
Fixed it